Add SACA framework Op and SACA use case

elevans opened the pull request:
This PR adds the SACA framework (3 Ops: z-score heat map, p-value heatmap, sig mask) and the use case that goes with it.

elevans committed:
Add SACA from imagej-ops dev branch

This commit adds Shulei and Ellen's SACA code from
imagej/imagej-ops coloc-multithread-saca branch
from commit b12ce2904945f7cf14db105e86f917d113279b63. This is
right before Ellen commited a WIP first attempt to multithread SACA.
While it works there are notes indicating that the output is still slow
and does not conform to Shulei's output from the original R package.
Given this, I felt it best to start the SciJava Ops port from the
last known working state and then make progress towards resolving
the Op performance and output results in the SciJava Ops framework.

elevans committed:
Convert SACA ImageJ Op to SciJava Op

This commit converts the imagej op copied from
the saca branch to a SciJava op. This Op works, however
it is not multithreaded and it produces NaN results at the
edge of the heatmap.

elevans committed:
Add check for NaN results and set to 0

Shulei's original SACA code also sets NaN's to 0.

elevans committed:
Use multithreaded Loopbuilder for SACA

elevans committed:
Use chunking with LoopBuilder in singleiteration

To multithread the singleiteration function, we need
to use the chunked LoopBuilder. This results in about
a ~40% reducing in processing time.

gselzer committed:
Add RAI focusers

elevans committed:
Rename saca op to saca.heatmapZScore

The SACA framework has three components:

- Z score heatmap
- SigPixel matrix
- pValue matrix

The Z score heatmap is the result of weighted kendall tau
operation. The SigPixel (significant pixel) and pValue results
are derived from the heatmap Z score. Thus to use the SigPixel
and pValue matrix outputs the user needs to first create
the Z score heatmap.

elevans committed:
Add QNorm class from imagej-ops saca branch

elevans committed:
Convert SACA op from a Function to Computer

The goal here is to let Adaptation make this op
a Function.

elevans committed:
Add SACA significant pixel mask output

elevans committed:
Add SACA PValue heatmatp op

elevans committed:
Clean up imports and improve SACA Op descriptions

elevans committed:
Rebase with main -- remove arity() syntax

This commit is a rebase of the saca branch with main.
For some reason the simplify/RAIFocusersTest.java file
was kept and not removed during the rebase. This commit
removes it and updates the OpRegressionTest count.

elevans committed:
Use non-multithreaded loop builder

The loop builder introduces a non-reproducible
randomness to the SACA algorithm. Unfortnately,
we need stop using the multithreaded version of the
loop builder to resolve this.

elevans committed:
Limit the SACA heatmap z score op to 2D

elevans committed:
Apply maven formatter, make it happy

elevans committed:
Add test for SACA framework

This commit adds a test for the SACA framework.
It tests all three Ops in the "coloc.saca" namespace.
A slice is taken of both input test data and processed
using SACA. 10 positions are sampled on all three SACA
outputs.

elevans committed:
Add SACA use case

gselzer commented on scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.vm:
```diff
@@ -0,0 +1,101 @@
+/*
+ * #%L
```
Can we delete this?

elevans replied:
Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

gselzer commented on scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.vm:
```diff
@@ -0,0 +1,72 @@
+/*
+ * #%L
+ * ImageJ2 software for multidimensional image processing and analysis.
```
Can we delete this?

elevans replied:
Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

gselzer commented on scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.list:
```diff
@@ -0,0 +1,25 @@
+.include templates/main/java/org/scijava/ops/image/adapt/Globals.list
+
+[RAIFocusers.java]
```
Can we delete this?

elevans replied:
Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

gselzer commented on scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.list:
````diff
@@ -0,0 +1,23 @@
+
+[RAIFocusersTest.java]
+
+types = ```
````
Can we delete this?

elevans replied:
Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

gselzer commented on scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -670,10 +670,17 @@ private List<RichOp<?>> resolveOpDependencies(OpInfo info,
 	private OpRequest inferOpRequest(OpDependencyMember<?> dependency,
 		Map<TypeVariable<?>, Type> typeVarAssigns)
 	{
```
Is this change necessary?

elevans replied:
No it's not. I tried looking to see where these changes came from but its a bit murky. My best guess is that I rebased from an experimental `scijava-ops-engine` branch that you were still working on and I just carried it over. Anyways, we definitely don't need this. Fixed here: aae10fb4c6e1edd09da852d3b9208cd2bc09323a.

gselzer commented on scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -670,10 +670,17 @@ private List<RichOp<?>> resolveOpDependencies(OpInfo info,
+		final Map<TypeVariable<?>, Type> inferences = new HashMap<>(typeVarAssigns);
+		for (var entry : inferences.entrySet()) {
+			var value = entry.getValue();
+			if (value.equals(Any.class) || value instanceof Any) {
```
If these changes *are* necessary, I wrote `Any.is()` to simplify this check

elevans replied:
The changes are necessary. Fixed here: aae10fb4c6e1edd09da852d3b9208cd2bc09323a

gselzer commented on scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/adapt/AdaptationMatchingRoutine.java:
```diff
@@ -293,7 +293,7 @@ private OpRequest inferOpRequest(Type type, String name,
 	@Override
 	public double priority() {
-		return Priority.LOW;
+		return Priority.VERY_LOW;
```
Is this change necessary?

elevans replied:
No it's not. `Priority.LOW` works just fine! Fixed: aae10fb4c6e1edd09da852d3b9208cd2bc09323a

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/AdaptiveSmoothedKendallTau.java:
```diff
@@ -0,0 +1,325 @@
+
+	private AdaptiveSmoothedKendallTau() {}
+
+	// TODO: check that output float type is actually what we want here.
```
Is this `TODO` still valid?

elevans replied:
It is not, good catch. I checked the other `TODO`s there is only 1 left that might need a look at later down the road (swaping Tau values instead of copying them). Fixed in 72d2d7dc14cc0691b67d23aca3089df7aad10bf1

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/PNorm.java:
```diff
@@ -0,0 +1,65 @@
+
+	private PNorm() {}
+
+	public static void compute(RandomAccessibleInterval<DoubleType> input,
```
I feel like this could be its own Op

elevans replied:
Agreed! It's now its own Op in the `stats` namespace! 72d2d7dc14cc0691b67d23aca3089df7aad10bf1

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/QNorm.java:
```diff
@@ -0,0 +1,67 @@
+ * @author Curtis Rueden
+ */
+
+public final class QNorm {
```
These could also be simple Op methods

elevans replied:
Agreed! `QNorm` is now an Op in the `stats` namesapce! fefe8bfb2a558ee1bc4a1306963d97e184206e99

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/SACAHeatmapPValue.java:
```diff
@@ -0,0 +1,73 @@
+ * @implNote op names='coloc.saca.heatmapPValue', priority='100.'
+ */
+
+public class SACAHeatmapPValue implements
```
This is fine as is, but if you didn't want to implement the `Computers.Arity2` interface you could easily make it a static method.

hinerm replied:
I know we discussed today but wanted to note that this class will be deleted in favor of the opified PNorm

elevans replied:
Correct! This class has been deleted.

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/WtKendallTau.java:
```diff
@@ -0,0 +1,236 @@
+ * @author Curtis Rueden
+ */
+
+final class WtKendallTau {
```
Could also consider making these their own ops

gselzer replied:
This one I feel less strongly about, because I do not know whether these would be used elsewhere

elevans replied:
I'm against making the `WtKendallTau.java` helper class an Op. It calculates the weighted [`kendall tau rank correlation`](https://en.wikipedia.org/wiki/Kendall_rank_correlation_coefficient) that determines the level/strength of colocalization within a set of data. My understanding here is that this kendall tau class is a specific implementation for SACA and would require a little bit of work to generalize. For now I think this is okay as is.

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/simplify/RAIFocusers.java:
```diff
@@ -0,0 +1,227 @@
+// * Redistribution and use in source and binary forms, with or without
+// * modification, are permitted provided that the following conditions are met:
+// *
+// * 1. Redistributions of source code must retain the above copyright notice,
```
I'd delete this

elevans replied:
Done! 6c9419aa8c9ad58f73e5e40d271c980cae2d1da8

gselzer commented on scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java:
```diff
@@ -0,0 +1,108 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Tests {@code coloc.saca} ops.
```
Add yourself as an author

elevans replied:
Done!

gselzer commented on scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java:
```diff
@@ -0,0 +1,108 @@
+public class SACATest extends AbstractColocalisationTest {
+
+	@Test
+	public void testSACAFramework() {
```
I'd split these into 3 separate tests for granularity and readability. i.e. it's nice to see quickly which of the Ops fail if any fail.

elevans replied:
The test has been split into 2 now! The `coloc.saca.heatmapPValue` Op no longer exists. Its replaced by the `stats.prnom` Op, which has its own test now.

gselzer reviewed (changes requested):
@elevans by and large this looks great! Thank you for porting this - just a couple small changes, none of which I'd expect to take any real work.

The one strange thing here is that there are edits to `RAIFocusers.java` and related templates which are probably unnecessary. Can we see if this PR still works without those changes? Same goes for the edits to SciJava Ops Engine changes - I think I put comments on each

elevans commented:
@gselzer Absolutely! So the `RAIFocusers` commit came from a rebase that had merge conflicts. I probably messed this up. So if this works without these changes, Great!!

elevans committed:
Remove simply/RAIFocusers

I accidently left these RAIFocusers in from
a rebase.

elevans committed:
Remove old/experimental changes to ops engine

This commit removes some code changes that I think
I applied from a rebase from one experimental branch
to the SACA one. I checked the log history in main and
couldn't find the code changes there. Regardless, they are not
needed. This commit should now align with the current ops
engine in main.

elevans committed:
Remove resolved TODOs in kendelltau helper class

elevans committed:
Convert SACA PNorm helper class into an Op

This commit converts the PNorm.java helper class
into an Op in the `stats` namespace. This commit also
adds a test for this new Op.

hinerm commented:
I don't see anything that needs revision beyond what Gabe said. Thanks @elevans!

elevans commented:
Thanks! Almost done!

elevans committed:
Convert SACA QNorm helper class into an Op

This commit converts the QNorm.java helper class
for SACA into an Op. This commit also includes a test
for this new Op.

elevans committed:
Split SACA framework test into two

elevans committed:
Update saca use case to use "stats.pnrom"

elevans committed:
Fix SACA tests

This commit refactors the coloc.saca namespace tests

elevans commented:
@gselzer Thanks for the awesome review. I think I've replied to everything and addressed all comments. Its ready for another look!

elevans committed:
Enable task progress for coloc.saca.heatmapZScore

gselzer commented on docs/ops/doc/examples/saca.rst:
```diff
@@ -0,0 +1,167 @@
+
+You can download the colocalization dataset `here`_.
+
+.. figure:: https://media.imagej.net/scijava-ops/1.0.0/saca_input.png
```
With #158 we should rebase this branch over the latest `main`, and update the `media.imagej.net` links to `media.scijava.org`

elevans replied:
Done!

gselzer commented on docs/ops/doc/examples/saca.rst:
```diff
@@ -0,0 +1,167 @@
+        
+        # print Pearson's and Li's results
+        print("pearsons: " + str(pearsons))
+        print("li: " + str(li))
```
It'd be nice to display the expected values of these returns

elevans replied:
Done! I agree, this was an over sight.

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/stats/DefaultPNorm.java:
```diff
@@ -0,0 +1,79 @@
+
+public final class DefaultPNorm implements
+	Computers.Arity2<RandomAccessibleInterval<DoubleType>, Boolean, RandomAccessibleInterval<DoubleType>>
+{
```
Any reason why this has to be a `RAI<DoubleType>`? Could it be a `RAI<O extends RealType<O>>`?

elevans replied:
There is not reason and in fact with your help we made this Op better be allowing it work on pixels instead of just RAIs!

gselzer commented on scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java:
```diff
@@ -0,0 +1,123 @@
+
+		// get random access and assert results are equal
+		final RandomAccess<BitType> sRA = sigMask.randomAccess();
+		for (int i = 0; i < xPositions.length; i++) {
```
I often forget about this, but there's a nice method `AbstractOpTest.areCongruent()` to check image equality and also `AbstractOpTest.asArray(Iterable<T> image)` that you could use to avoid manually iterating over the output image.

You don't have to do anything, just an FYI, and also a reminder for me to use it!

elevans replied:
This would be awesome to use, but unfortunately for me I'm extending `AbstractColocalisationTest` and not `AbstractOpTest` where `areCongruent()` lives. I'm just going to stick with this manual way for now :D.

gselzer commented on scijava-ops-image/src/test/java/org/scijava/ops/image/stats/DefaultPNormTest.java:
```diff
@@ -0,0 +1,80 @@
+			0.999998128463855 };
+
+		// load Z-score heatmap slice
+		Img<FloatType> zscore = openFloatImg("zscore_test_data.tif");
```
Nothing in the algorithm depends upon the data distribution in the image, correct? If so, then could we:
* Save the 10 data points in a final `double[]`
* Create an `ArrayImg` from that `double[]` using `ArrayImgs.doubles(data, 10)`
* Continue as normal (except you'd have one fewer dimension in your image, which is an added bonus 😉)
That way, we wouldn't have to ship an image with the tests.

elevans replied:
Done! With this change I was able to remove the need to read an image and also ship an image in the tests!

gselzer reviewed (changes requested):
@elevans thank you so much for addressing my prior review! More Ops = a better framework!

...unfortunately, I'd suggest a couple more things 😅

elevans committed:
Refactor PNorm Op into a pixel-wise Op

elevans committed:
Update SACA example URLs and add coloc values

elevans committed:
Refactor Abstract coloc test to use static methods

elevans committed:
Refactor SACATest to use setup helper method

This commit refactors the SACA test to use a setup helper
method create the resources needed for the tests just once.

elevans committed:
Refactor PNorm test to use ArrayImgs

elevans commented:
@gselzer This PR is ready for its second round of review. Thanks for all the help!
